.github/workflows/haskell-ci.yml

```diff
@@ -175,37 +175,52 @@ jobs:
         else
           echo "PROJECT_ROOT=${WORK_DIR}" >> $GITHUB_ENV
         fi
-    # TODO(#19): Separate caches for build directory and Cabal store
-    - name: Restore Cabal cache
+    - name: Post-GHC installation fixups on Windows
+      if: runner.os == 'Windows'
+      run: |
+        # A workaround for https://github.com/Mistuke/CabalChoco/issues/5
+        cabal user-config update -a "extra-include-dirs: \"\""
+        cabal user-config update -a "extra-lib-dirs: \"\""
+    - name: Configure
+      run: cabal configure ${{ inputs.configure-flags }} -- ${{ inputs.build-targets }} ${{ inputs.test-targets }}
+      working-directory: ${{ env.WORK_DIR }}
+    # Creates `dist-newstyle/cache/plan.json` for the cache key.
+    - name: Create build plan
+      run: cabal build --dry-run -- ${{ inputs.build-targets }}
+      working-directory: ${{ env.WORK_DIR }}
+    - name: Restore Cabal store cache
       uses: actions/cache/restore@v4
-      id: cache
+      id: store-cache
       env:
         # Each `inputs.os` (e.g., `ubuntu-22.04-arm`) uniquely determines a
         # `runner.arch` (e.g., ARM64), so there is no need to include the latter
         # as part of the cache key.
-        key: ${{ inputs.cache-key-prefix }}-haskell-ci-${{ inputs.os }}-cabal${{ steps.setup-haskell.outputs.cabal-version }}-ghc${{ steps.setup-haskell.outputs.ghc-version }}
+        key: ${{ inputs.cache-key-prefix }}-haskell-ci-cabal-store-${{ inputs.os }}-cabal${{ steps.setup-haskell.outputs.cabal-version }}-ghc${{ steps.setup-haskell.outputs.ghc-version }}
+      with:
+        path: ${{ steps.setup-haskell.outputs.cabal-store }}
+        # The `restore-key` allows for restoring the cache from a different
+        # plan, which can still speed up the build if most dependencies were
+        # unchanged.
+        key: |
+          ${{ env.key }}-${{ hashFiles(format('{0}/dist-newstyle/cache/plan.json', env.PROJECT_ROOT)) }}
```

Comment:
Can you leave comments about what this `plan.json` file is and why we include it in the cache key?